### PR TITLE
Uninteresting change: introducing `ClientApplicationBase`

### PR DESCRIPTION
**Pull request by nikitamikhaylov**
A part of this PR: https://github.com/ClickHouse/ClickHouse/pull/48902. Trying to extract the part of the diff which is pure Ctrl-C + Ctrl-V from one file to another, because it's huge and difficult to track changes coming from master branch.

### Changelog category (leave one):

- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [x]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit a93dae7da2bfc2f2da54dac64f82846b99759cb8 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/66549/a93dae7da2bfc2f2da54dac64f82846b99759cb8/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Client/ClientBase.cpp
Comment (jkartseva) [inline src/Client/ClientBase.cpp:1]: `#include <Core/BaseSettingsProgramOptions.h>` can be removed.

Also, I don't see where `using namespace std::literals;` is used in this file.

**Review by jkartseva — APPROVED** [review]

**Comment by nikitamikhaylov** [comment]
`Stateless tests (coverage) [3/5]` - https://github.com/ClickHouse/ClickHouse/issues/66589

File: src/Client/ClientBase.cpp
Comment (nikitamikhaylov) [reply src/Client/ClientBase.cpp:1]: I will fix it a part of other PR - don't want to rerun the CI.

File: src/Client/ClientApplicationBase.cpp [hunk 1]
```diff
@@ -0,0 +1,507 @@
...
+        return ClientInfo::QueryKind::SECONDARY_QUERY;
+    if (query_kind == "no_query")
+        return ClientInfo::QueryKind::NO_QUERY;
+    throw Exception(ErrorCodes::BAD_ARGUMENTS, "Unknown query kind {}", query_kind);
+}
+
+/// This signal handler is set only for SIGINT and SIGQUIT.
+void interruptSignalHandler(int signum)
+{
+    if (ClientApplicationBase::getInstance().tryStopQuery())
```
Comment (Algunenano) [inline src/Client/ClientApplicationBase.cpp:50]: This seems to be the cause of the https://github.com/ClickHouse/ClickHouse/issues/66589 crash

**Review by Algunenano — APPROVED** [review]

**Comment by nikitamikhaylov** [comment]
`Stateless tests (tsan)`: https://github.com/ClickHouse/ClickHouse/pull/66659
`Stateless tests (ubsan)`: https://github.com/ClickHouse/ClickHouse/issues/47866

Build with clang-tidy was broken in master.

**Pull request merged**